db/clientlist_virtual_reader.cc

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "db/clientlist_virtual_reader.hh"
+
+#include "api/api.hh"
+#include "db/system_keyspace.hh"
+#include "db/timeout_clock.hh"
+#include "dht/i_partitioner.hh"
+#include "mutation_fragment.hh"
+#include "range.hh"
+#include "service/storage_service.hh"
+#include "transport/server.hh"
+
+#include <algorithm>
+#include <unordered_map>
+#include <vector>
+
+namespace db::clientlist {
+
+namespace {
+class clients_reducer {
+    std::vector<system_keyspace::client_data> _val;
+public:
+    future<> operator()(std::vector<system_keyspace::client_data>&& rhs) {
+        // Move the smaller vector into the bigger one
+        if (_val.size() < rhs.size()) {
+            _val.swap(rhs);
+        }
+        _val.reserve(_val.size() + rhs.size());
+        _val.insert(_val.end(), std::make_move_iterator(rhs.begin()), std::make_move_iterator(rhs.end()));
+        return make_ready_future<>();
+    }
+
+    std::vector<system_keyspace::client_data> get() && {
+        return std::move(_val);
+    }
+};
+
+class ck_comparator {
+    schema_ptr _schema;
+public:
+    using port_type = decltype(system_keyspace::client_data::port);
+
+    ck_comparator(schema_ptr schema) : _schema(schema) {}
+
+    int operator()(const clustering_key_prefix& lhs, const clustering_key_prefix& rhs) {
+        return clustering_key_prefix::prefix_equal_tri_compare(*_schema)(lhs, rhs);
+    }
+    int operator()(const port_type& lhs, const clustering_key_prefix& rhs) {
+        const auto lhs_bytes = data_type_for<port_type>()->decompose(lhs);
+        return operator()(clustering_key_prefix::from_single_value(*_schema, lhs_bytes), rhs);
+    }
+    int operator()(const clustering_key_prefix& lhs, const port_type& rhs) {
+        const auto rhs_bytes = data_type_for<port_type>()->decompose(rhs);
+        return operator()(lhs, clustering_key_prefix::from_single_value(*_schema, rhs_bytes));
+    }
+};
+}  // anon. namespace
+
+static future<std::vector<system_keyspace::client_data>> get_all_clients_data(
+        distributed<cql_transport::cql_server>& distributed_cql_server) {
```

Comment:
It's easy for the connection table to be in the thousands per shard. Multiplying by 100 shards, that's hundreds of thousands of objects. That's at the verge of starting to cause problems (stalls/bad_allocs).

We should consider an incremental interface, with this function receiving a last key from which to start. Unfortunately it makes a lot harder and uglier, since cql_server will now need to understand the partition key of this virtual table and maintain connections in sorted order. @tgrabiec any ideas?

Reply:
Maybe we should store this in a non-virtual local table. It adds an extra write on connect/disconnect, but that's nothing, and it's a lot simpler. Also need to truncate on startup.

Reply:
I don't think we need a local table. For starts each shard should only emit connection info whose partition key it owns. We can filter out non-belonging connections early.

It would all be much simpler if we used a custom partitioner that sharded connections such that each shards own connections belonging to it. Then connections could be members of an intrusive set/map which takes care of ordering. Maybe at this point a local table is simpler though. Also I'm not sure how well our code is prepared for any partitioner other than the global one. Can the partition even be per table?

Reply:
We're not prepared per-table partitioners at all. Even though we only need to specialize shard_of().

Sure each shard will emit only owned connections, but how to do the merge sort? It's not impossible to solve, but also not easy.

A regular disk-based table solves this and doesn't occupy memory when not in use. A virtual table would be nicer, but needs a large amount of work to get right. Some of that work can be reused with other virtual tables if we manage to make it generic enough, but that's not easy either.